### ISSUE TITLE
Why is `^|\s+` used instead of just `\s+` in .split()?

### ISSUE THREAD

**Issue by ChALkeR** [d3/d3-dispatch#18]
On the following lines, `^|\s+` is used:
https://github.com/d3/d3-dispatch/blob/893b6f403c2e057208aea1bb30ba5623e2c38ccd/src/dispatch.js#L15-L17

Does that do anything different from what just `\s+` would do?

It causes problems when trying to use this library in QML environment (V4 JS engine from Qt).
Ref: [QTBUG-62471](https://bugreports.qt.io/browse/QTBUG-62471).

**Comment by mbostock** [comment]
I think the only difference is when *typenames* is the empty string or entirely whitespace.

```js
"".split(/^|\s+/) // []
"".split(/\s+/) // [""]
```

**Comment by ChALkeR** [comment]
@mbostock Ah, thanks, I totally missed that.

Perhaps that could be solved with `.filter(function(t) { return t; })` instead?
Or, a length check after `.trim()`.

**Comment by ChALkeR** [comment]
I'm not even sure though that this would make things work under Qt QML, I checked pretty long ago.
Perhaps that needs a recheck before anything is done here.

**Comment by mbostock** [comment]
I’m not planning on taking any action. Thank you for the question.

**Comment by localpcguy** [comment]
I think I've found a (browser?) bug related to this regex and the question asked by @ChALkeR here. In Internet Explorer (v11 is where I saw this), the regex `/^|\s+/` splits a string into it's individual characters. So something like `"end".trim().split(/^|\s+/)` in IE11 outputs `["e", "n", "d"]`. In modern browsers, it outputs `[ "end" ]` as expected.

The issues seems to be the pipe, it looks like it causes IE to split not on the whitespace but on nothing, effectively. 

I'm not sure what the expected browser support is, but as suggested above, using `.split(/\s+/).filter(function(t) { return t; })` would achieve the same result, work in IE9+ (when `.filter` was added), and eliminate the regex bug. 

Let me know if this should be a new bug.
